src/script.c

```diff
@@ -108,10 +108,62 @@ int scriptInterrupt(scriptRunCtx *run_ctx) {
 }
 
 /* Prepare the given run ctx for execution */
-void scriptPrepareForRun(scriptRunCtx *run_ctx, client *engine_client, client *caller, const char *funcname) {
+int scriptPrepareForRun(scriptRunCtx *run_ctx, client *engine_client, client *caller, const char *funcname, uint64_t function_flags, int ro) {
     serverAssert(!curr_run_ctx);
-    /* set the curr_run_ctx so we can use it to kill the script if needed */
-    curr_run_ctx = run_ctx;
+
+    if (!(function_flags & SCRIPT_FLAG_IGNORE_FLAGS)) {
+        if ((function_flags & SCRIPT_FLAG_NO_CLUSTER) && server.cluster_enabled) {
+            addReplyError(caller, "Can not run function on cluster, 'no-cluster' flag is set.");
+            return C_ERR;
+        }
+
+        if (!(function_flags & SCRIPT_FLAG_ALLOW_OOM) && server.script_oom && server.maxmemory) {
+            addReplyError(caller, "-OOM allow-oom flag is not set on the function, "
+                                  "can not run it when used memory > 'maxmemory'");
+            return C_ERR;
+        }
+
+        if (server.masterhost && server.repl_state != REPL_STATE_CONNECTED &&
+            server.repl_serve_stale_data == 0 && !(function_flags & SCRIPT_FLAG_ALLOW_STALE))
+        {
+            addReplyError(caller, "-MASTERDOWN Link with MASTER is down, "
+                             "replica-serve-stale-data is set to 'no' "
+                             "and 'allow-stale' flag is not set on the function.");
+            return C_ERR;
+        }
+
+        if (!(function_flags & SCRIPT_FLAG_NO_WRITES)) {
+            /* Function may perform writes we need to verify:
+             * 1. we are not a readonly replica
+             * 2. no disk error detected
+             * 3. command is not 'fcall_ro' */
+            if (server.masterhost && server.repl_slave_ro && caller->id != CLIENT_ID_AOF
+                && !(caller->flags & CLIENT_MASTER))
+            {
+                addReplyError(caller, "Can not run a function with write flag on readonly replica");
```

Comment:
We need to generalise error message to specify scripts and not functions

Reply:
:+1: